### ISSUE TITLE
Not quite an issue, just a question

### ISSUE THREAD

**Issue by ovizii** [fmbla/spamassassin-concepts#1]
I think I have understood how this works but I am unsure about how it interacts with BAYES. 
You are saying:

> Usage
> 
> In day to day life you do not need to interact with Concepts - Bayes will automatically keep up to >date with what is considered good and bad as long as you classify your spam and ham carefully.

Does that mean that basically as long as the concepts extracted are fitting the emails well and I teach BAYES SPAM/HAM it will pickup which concepts are used in SPAM/HAM and thus is benefitting from the concepts? Basically injecting keywords for BAYES use?

**Comment by steadramon** [comment]
Almost. It adds the found concepts to a header within the email - X-SA-Concepts. This header can be used by SA bayes to add tokens to the Bayes DB. The concepts go into the Bayes DB as unique tokens, so the "hello" concept differs to the standard "hello" token appearing in the body.

More information about this can be found here: http://wiki.junkemailfilter.com/index.php/Concept_Parsing_Spam_Filter

**Comment by ovizii** [comment]
Thanks for getting back to me but you managed to confuse me even more :-) -  I had read that link before and after reading this on that page:

> You could write combination rules or feed these tokens into Bayes to make it self scoring. 

And on your own page this:

> In day to day life you do not need to interact with Concepts - Bayes will automatically keep up to >date with what is considered good and bad as long as you classify your spam and ham carefully.

I understand that  BAYES is using those concepts from the X-SA-Concepts header automatically and yet in this reply, you are saying that:

> This header can be used by SA bayes to add tokens to the Bayes DB.

So is this automated or do I have to manually setup "filters" like the examples you gave:

> ## JUST AN EXAMPLE!!!
> 
> header __CONCEPTS_GOD X-SA-Concepts =~ /\bgod\b/
> header __CONCEPTS_AFRICA X-SA-Concepts =~ /\bafrica\b/
> header __CONCEPTS_LOTSMONEY X-SA-Concepts =~ /\blotsofmoney\b/
> header __CONCEPTS_STRANGER X-SA-Concepts =~ /\bstranger\b/
> header __CONCEPTS_DYING X-SA-Concepts =~ /\bdying\b/
> 
> #scores 4 or more
> meta GOD_IN_AFRICA_LOTSAMONEY_STRANGERS_DYING (__CONCEPTS_GOD + >__CONCEPTS_AFRICA + __CONCEPTS_LOTSMONEY + __CONCEPTS_STRANGER + >__CONCEPTS_DYING >= 4)
> describe GOD_IN_AFRICA_LOTSAMONEY_STRANGERS_DYING Combo
> score GOD_IN_AFRICA_LOTSAMONEY_STRANGERS_DYING 1.0

**Comment by steadramon** [comment]
The Bayes interaction is automated - I am not the author of the page above, however did get the author's permission to implement this in SA.

The meta rules addition allows you to manually score combinations of Concept hits.

**Comment by ovizii** [comment]
Perfectly clear now. 
So for initial tests, I just let it run without any extra meta rules and BAYES will pickup the tokens added by this plugin. 

Which again leads me to the question, how/why does BAYES treax tokens differently from the X-SA-Concepts header than tokens from within an email body?

**Comment by steadramon** [comment]
That's right.

Tokens are hashed before being sent to the DB for storage. Tokens from headers are prefixed with the header name, so in this case our token is actually a hash of 

`H$X-SA-Concepts:hello`

rather than just 'hello'

Multiple Concepts = multiple tokens

**Comment by ovizii** [comment]
hm, sounds like an intriguing concept, pun intended. 

we could actually add another language into the same concepts. 
say take the concept hello, translate all the content into another language but the concept file name remains hello. 

I need to play with this some more :-)

**Comment by steadramon** [comment]
Yes! That's the general idea - the concepts files could take a lot of looking after, however I think it has it's benefits.

There was a discussion on the SA mailing list about this - http://www.gossamer-threads.com/lists/spamassassin/users/198896

**Comment by ovizii** [comment]
I read that discussion and was ready to give it a go. I copied everything into place and waited for an email to come through. 

If I manually check this email:
````
spamassassin -D -t /var/spool/MailScanner/quarantine/20161129/nonspam/0ECC910002A.AD2AC 2>&1 | grep Concepts
Nov 29 13:20:20.595 [15197] dbg: config: read file /etc/mail/spamassassin/Concepts.cf
Nov 29 13:20:21.466 [15197] dbg: config: fixed relative path: /etc/mail/spamassassin/Concepts.pm
Nov 29 13:20:21.466 [15197] dbg: plugin: loading Mail::SpamAssassin::Plugin::Concepts from /etc/mail/spamassassin/Concepts.pm
Nov 29 13:20:22.005 [15197] dbg: plugin: Mail::SpamAssassin::Plugin::Concepts=HASH(0x379c550) implements 'finish_parsing_start', priority 0
Nov 29 13:20:22.020 [15197] dbg: Concepts: 273 concepts loaded
Nov 29 13:20:22.699 [15197] dbg: plugin: Mail::SpamAssassin::Plugin::Concepts=HASH(0x379c550) implements 'extract_metadata', priority 0
Nov 29 13:20:22.938 [15197] dbg: Concepts: metadata: X-SA-Concepts: disclaimer https secret search-eng internet reply2me twitter forwarded contact facebook stranger time-ref please regards friend phone-num email-adr
Nov 29 13:20:22.939 [15197] dbg: plugin: Mail::SpamAssassin::Plugin::Concepts=HASH(0x379c550) implements 'parsed_metadata', priority 0
````

I can see the X-SA-Concepts metadata but if I look at my email: /var/spool/MailScanner/quarantine/20161129/nonspam/0ECC910002A.AD2AC I can't see any headers containing the word "concepts" - was I wrong in assuming I would see all the concepts which were applied to this email in a header?
